File: Array/easy/1512. Number of Good Pairs/numIdenticalPairs.py
#time complexity : O(n^2)
#space complexity : O(1)
class Solution(object):
    #in this solution i walk through the list with nested loops and find our branch
    def numIdenticalPairs(self, nums):
        n = len(nums)
        k = 0
        for i in range(n):
            for j in range(n):
                if i < j and nums[i] == nums[j]:
                    k += 1
                    
        return k
    
#time complexity : O(n)
#space complexity : O(n)
class Solution(object):
    #in this solution i walk through list with a loop then add each element in to dict that values are count of each element in the list
    def numIdenticalPairs(self, nums):
        dictionary = {}
        k = 0
        for i in nums:
            #when meet same key, plus it's value to k then plus one to it's value 
            if i in dictionary:
                k += dictionary[i]
                dictionary[i] += 1
            #if not in dict add it and set it's count to 1
            else:
                dictionary[i] = 1
            
        return k

File: Array/easy/1512. Number of Good Pairs/numIdenticalPairs.cpp
//time complexity : O(n^2)
//space complexity : O(1)
class Solution 
{
public:
    //in this solution i walk through the vector with nested loops and find our branch
    int numIdenticalPairs(vector<int>& nums)
    {
        int len = nums.size();
        int k = 0;
        for (int i = 0; i < len; i++)
        {
            for (int j = 0; j < len; j++)
            {
                if (nums[i] == nums[j] and i < j) k++;
            }
        }
        return k;
    }
};

//time complexity : O(n)
//space complexity : O(n)
class Solution 
{
public:
    //in this solution i walk through vector with a loop then add each element in to hash map that values are count of each element in the vector
    int numIdenticalPairs(vector<int>& nums)
    {
        unordered_map<int, int> map;
        int k = 0;
        for (int i = 0; i < nums.size(); i++)
        {
            //when meet same key, plus it's value to k then plus one to it's value 
            if (map[nums[i]])
            {
                k += map[nums[i]];
                map[nums[i]] += 1;
            }
            //if not in dict add it and set it's count to 1
            else
            {
                map[nums[i]] = 1;
            }
        }
        return k;
    }
};